File: Closure/7 kyu - Complementary DNA.clj
; https://www.codewars.com/kata/554e4a2f232cdd87d9000038/train/clojure
; My solution
(ns complementary-dna)

(defn dna-compl [base]
  (if (= base "A") "T"
  (if (= base "T") "A"
  (if (= base "G") "C"
  (if (= base "C") "G" "")))))

(defn dna-strand [dna]
  (apply str
    (map (comp dna-compl str) dna)))

; ...
(ns complementary-dna)

(defn dna-strand [dna]
    (apply str (map {\A \T \C \G \G \C \T \A} dna)))

; ...
(ns complementary-dna)

(require '[clojure.string :refer [replace]])

(defn dna-strand [dna]
  (replace dna #"\S" {"A" "T" "C" "G" "T" "A" "G" "C"}))

; ...
(ns complementary-dna)

(defn dna-strand [dna]
  (apply str (map #(case % \A \T \T \A \C \G \G \C) dna)))

; ...
(ns complementary-dna
  (:require [clojure.string :as string]))

(defn dna-strand [n]
  (string/join (replace {"A" "T", "T" "A", "C" "G", "G", "C"}
                        (string/split n #""))))

; ...

File: TypeScript/7 kyu - Complementary DNA.ts
// https://www.codewars.com/kata/complementary-dna/train/typescript
// My solution
export class Kata {
    static dnaStrand(dna: string) {
        const map = { "A": "T", "T": "A", "C": "G", "G": "C" };
        return dna.split("").map(symbol => map[symbol]).join("");
    }
}
// ...
export class Kata {
    static dnaStrand(dna: string) {
        return dna.replace(/./g, (c) => ({ A: 'T', T: 'A', G: 'C', C: 'G' })[c]);
    }
}
// ...
interface IDictionary {
    [index: string]: string;
}

var params: IDictionary = {};
params['A'] = 'T';
params['T'] = 'A';
params['G'] = 'C';
params['C'] = 'G';

export class Kata {
    static dnaStrand(dna: string) {
        let result: string = "";
        let arrayLength: number = dna.length;
        let i: number = 0;

        for (i = 0; i < arrayLength; i++) {
            result = result + params[dna[i]];
        }

        return result;
    }
}
// ...
export class Kata {
    static dnaStrand(dna: string): string {
        const match = {
            'A': 'T',
            'T': 'A',
            'G': 'C',
            'C': 'G'
        }
        return Array.from(dna, v => match[v]).join('');
    }
}
// ...
export class Kata {
    static dnaStrand(dna: string) {
        return dna.split('').map(c => map.get(c)).join('');
    }
}

export const map: Map<string, string> = new Map<string, string>()
    .set('A', 'T')
    .set('C', 'G')
    .set('G', 'C')
    .set('T', 'A');
// ...
export class Kata {
    static dnaStrand(dna: string) {
        let complement = { "A": "T", "T": "A", "C": "G", "G": "C" };
        return dna.replace(/[atcg]/gi, m => complement[m]);
    }
}
// ...
export class Kata {
    static dnaStrand(dna: string) {
        //your code here
        const CODES = { A: 'T', T: 'A', C: 'G', G: 'C' };
        return dna.split('').map(n => CODES[n]).join('');
    }
}
